node/src/Pos/Update/Download.hs

```diff
@@ -53,22 +53,34 @@ downloadUpdate :: forall ctx m . UpdateMode ctx m => ConfirmedProposalState -> m
 downloadUpdate cst@ConfirmedProposalState {..} = do
     unlessM (liftIO . doesFileExist =<< views (lensOf @UpdateParams) upUpdatePath) $ do
         downSetVar <- views (lensOf @UpdateContext) ucDownloadingUpdates
-        let upId = hash cpsUpdateProposal
-        whenM (tryPutToSet downSetVar upId) $
-            downloadUpdateDo cst
-            `finally` (atomically $ modifyTVar' downSetVar (S.delete upId))
+        (forDownload, downSet) <- tryPutToSet downSetVar cpsUpdateProposal
+        if forDownload
+           then do
+              logDebug $ sformat ("Update downloading triggered, download state: "%listJson)
+                                 downSet
+              downloadUpdateDo cst
+                `finally` clear downSetVar cpsUpdateProposal
+           else
+              logDebug $ sformat ("Update downloading already in progress, download state: "%listJson)
+                                 downSet
+
   where
+    clear downSetVar up = do
+        downSet <- atomically $ do
+            modifyTVar' downSetVar (HS.delete cpsUpdateProposal)
+            readTVar downSetVar
+        logDebug $ sformat ("Update "%build%" downloaded, download state: "%listJson)
```

Comment:
1. It seems that this code will be called when `downloadUpdateDo` throws an exception, so this debug messages is a bit misleading.
2. Passing `UpdateProposal` to the functions inside `where` is redundant in this case, because it's always `cpsUpdateProposal` and it's available to this scope. Or if you really want to pass it, it makes more sense to use `up` instead of `cpsUpdateProposal` two lines above.

Anyway, I am going to approve, because apparently I need to touch this code in CSL-1592 anyway.